File: compare/generate-report.js
var fs = require('fs');
var path = require('path');

var resultsPath = process.argv[2];
if (!resultsPath) {
  console.error('Usage: node generate-report.js <results.json>');
  process.exit(1);
}

var results = JSON.parse(fs.readFileSync(resultsPath, 'utf-8'));

var lines = [];
lines.push('# 西安地铁导航对比结果 — 全部案例');
lines.push('');
lines.push('> 运行时间: ' + (results[0] && results[0].id ? '2026-05-20' : 'N/A'));
lines.push('> 有效对比: ' + results.filter(function(r) { return r.comparison && r.comparison.local_ok && r.comparison.amap_ok; }).length + ' / ' + results.length + ' 组');
lines.push('');
lines.push('---');
lines.push('');

results.forEach(function(r) {
  var c = r.comparison;
  var local = r.local;
  var amap = r.amap;

  lines.push('## ' + r.id + ' ' + r.origin + ' → ' + r.dest + ' (' + (r.type === 'manual' ? '手工' : '随机') + ')');
  lines.push('');

  if (!c) {
    lines.push('> 对比数据缺失');
    lines.push('');
    return;
  }

  if (c.local_error && c.amap_error) {
    lines.push('> 本地错误: ' + c.local_error);
    lines.push('> 高德错误: ' + c.amap_error);
    lines.push('');
    return;
  }

  lines.push('| 指标 | 本地系统 | 高德地图 | 差异 |');
  lines.push('|------|---------|---------|------|');

  if (c.local_ok && c.amap_ok) {
    lines.push('| 总耗时(min) | ' + c.local_time_min.toFixed(2) + ' | ' + c.amap_time_min.toFixed(2) + ' | ' + c.time_diff_min.toFixed(2) + ' (' + c.time_diff_pct.toFixed(1) + '%) |');
    lines.push('| 换乘次数 | ' + c.local_transfers + ' | ' + c.amap_transfers + ' | ' + (c.transfer_diff > 0 ? '+' : '') + c.transfer_diff + ' |');
    lines.push('| 途经站数 | ' + c.local_station_count + ' | ' + c.amap_station_count + ' | ' + (c.station_diff > 0 ? '+' : '') + c.station_diff + ' |');
    lines.push('| 步行距离(m) | 0 | ' + c.amap_walking_m + ' | -' + c.amap_walking_m + ' |');
    lines.push('| Jaccard一致率 | - | - | ' + c.jaccard.toFixed(4) + ' |');
    lines.push('| 线路一致 | - | - | ' + (c.line_comparison && c.line_comparison.lines_match ? '✓ 是' : '✗ 否') + ' |');
  } else if (c.local_ok) {
    lines.push('| 总耗时(min) | ' + c.local_time_min.toFixed(2) + ' | N/A | - |');
    lines.push('| 换乘次数 | ' + c.local_transfers + ' | N/A | - |');
    lines.push('| 途经站数 | ' + c.local_station_count + ' | N/A | - |');
    lines.push('| 高德错误 | - | ' + (c.amap_error || '未知') + ' | - |');
  } else if (c.amap_ok) {
    lines.push('| 总耗时(min) | N/A | ' + c.amap_time_min.toFixed(2) + ' | - |');
    lines.push('| 换乘次数 | N/A | ' + c.amap_transfers + ' | - |');
    lines.push('| 本地错误 | ' + (c.local_error || '未知') + ' | - | - |');
  }

  lines.push('');

  if (local && !local.error && local.segments) {
    lines.push('**本地路线:**');
    lines.push('');
    local.segments.forEach(function(seg) {
      lines.push('- **' + seg.line + '**: ' + seg.from + ' → ' + seg.to + ' (' + seg.stations.length + '站)');
      if (seg.stations.length <= 10) {
        lines.push('  - ' + seg.stations.join(' → '));
      } else {
        lines.push('  - ' + seg.stations.slice(0, 5).join(' → ') + ' → ... → ' + seg.stations.slice(-3).join(' → '));
      }
    });
    if (local.transfer_stations && local.transfer_stations.length > 0) {
      lines.push('- 换乘站: ' + local.transfer_stations.join('、'));
    }
    lines.push('');
  }

  if (amap && !amap.error && amap.schemes && amap.schemes.length > 0) {
    var scheme = amap.schemes[0];
    lines.push('**高德路线:**');
    lines.push('');
    if (scheme.metro_segments && scheme.metro_segments.length > 0) {
      scheme.metro_segments.forEach(function(seg) {
        var viaCount = seg.via_stops ? seg.via_stops.length : 0;
        lines.push('- **' + seg.line_short + '**: ' + seg.departure_stop + ' → ' + seg.arrival_stop + ' (' + (viaCount + 2) + '站)');
        var allStops = [seg.departure_stop].concat(seg.via_stops || [], [seg.arrival_stop]);
        if (allStops.length <= 10) {
          lines.push('  - ' + allStops.join(' → '));
        } else {
          lines.push('  - ' + allStops.slice(0, 5).join(' → ') + ' → ... → ' + allStops.slice(-3).join(' → '));
        }
      });
      if (scheme.transfer_stations && scheme.transfer_stations.length > 0) {
        lines.push('- 换乘站: ' + scheme.transfer_stations.join('、'));
      }
    } else {
      lines.push('- 无纯地铁方案');
    }
    if (scheme.has_non_metro) {
      lines.push('- ⚠ 含非地铁段（公交/步行）');
    }
    lines.push('');
  }

  if (c && c.local_ok && c.amap_ok && c.transfer_comparison) {
    var tc = c.transfer_comparison;
    if (tc.local_only.length > 0 || tc.amap_only.length > 0) {
      lines.push('**换乘站差异:**');
      if (tc.common.length > 0) lines.push('- 共同换乘: ' + tc.common.join('、'));
      if (tc.local_only.length > 0) lines.push('- 仅本地: ' + tc.local_only.join('、'));
      if (tc.amap_only.length > 0) lines.push('- 仅高德: ' + tc.amap_only.join('、'));
      lines.push('');
    }
  }

  lines.push('---');
  lines.push('');
});

var outPath = path.join(path.dirname(resultsPath), 'full-report.md');
fs.writeFileSync(outPath, lines.join('\n'), 'utf-8');
console.log('Report written to: ' + outPath);

File: compare/generate_report.py
import json
import os
import sys


def build_lines(results):
    lines = ["# Xi'an Metro Compare Report", ""]
    valid_count = sum(1 for item in results if item.get("comparison", {}).get("local_ok") and item.get("comparison", {}).get("amap_ok"))
    lines.append(f"> Valid comparisons: {valid_count} / {len(results)}")
    lines.append("")
    lines.append("---")
    lines.append("")

    for item in results:
        compare = item.get("comparison") or {}
        local = item.get("local") or {}
        amap = item.get("amap") or {}

        lines.append(f"## {item['id']} {item['origin']} -> {item['dest']} ({item['type']})")
        lines.append("")

        if not compare:
            lines.append("> Missing comparison data")
            lines.append("")
            continue

        if compare.get("local_error") and compare.get("amap_error"):
            lines.append(f"> Local error: {compare['local_error']}")
            lines.append(f"> Amap error: {compare['amap_error']}")
            lines.append("")
            continue

        lines.append("| Metric | Local | Amap | Diff |")
        lines.append("|---|---:|---:|---:|")
        if compare.get("local_ok") and compare.get("amap_ok"):
            lines.append(f"| Time (min) | {compare['local_time_min']:.2f} | {compare['amap_time_min']:.2f} | {compare['time_diff_min']:.2f} ({compare['time_diff_pct']:.1f}%) |")
            lines.append(f"| Transfers | {compare['local_transfers']} | {compare['amap_transfers']} | {compare['transfer_diff']} |")
            lines.append(f"| Stations | {compare['local_station_count']} | {compare['amap_station_count']} | {compare['station_diff']} |")
            lines.append(f"| Walking (m) | 0 | {compare['amap_walking_m']} | -{compare['amap_walking_m']} |")
            lines.append(f"| Jaccard | - | - | {compare['jaccard']:.4f} |")
            lines.append(f"| Line match | - | - | {compare.get('line_comparison', {}).get('lines_match')} |")
        elif compare.get("local_ok"):
            lines.append(f"| Time (min) | {compare['local_time_min']:.2f} | N/A | - |")
            lines.append(f"| Local only | OK | {compare.get('amap_error', 'N/A')} | - |")
        else:
            lines.append(f"| Time (min) | N/A | {compare['amap_time_min']:.2f} | - |")
            lines.append(f"| Amap only | {compare.get('local_error', 'N/A')} | OK | - |")
        lines.append("")

        if local and not local.get("error") and local.get("segments"):
            lines.append("**Local route**")
            for segment in local["segments"]:
                lines.append(f"- **{segment['line']}**: {segment['from']} -> {segment['to']} ({len(segment['stations'])} stations)")
            if local.get("transfer_stations"):
                lines.append(f"- Transfers: {', '.join(local['transfer_stations'])}")
            lines.append("")

        if amap and not amap.get("error") and amap.get("schemes"):
            scheme = amap["schemes"][0]
            lines.append("**Amap route**")
            for segment in scheme.get("metro_segments", []):
                lines.append(f"- **{segment['line_short']}**: {segment['departure_stop']} -> {segment['arrival_stop']} ({len(segment.get('all_stops', []))} stations)")
            if scheme.get("transfer_stations"):
                lines.append(f"- Transfers: {', '.join(scheme['transfer_stations'])}")
            if scheme.get("has_non_metro"):
                lines.append("- Includes non-metro sections")
            lines.append("")

        transfer_cmp = compare.get("transfer_comparison")
        if transfer_cmp and (transfer_cmp["local_only"] or transfer_cmp["amap_only"]):
            lines.append("**Transfer station differences**")
            if transfer_cmp["common"]:
                lines.append(f"- Common: {', '.join(transfer_cmp['common'])}")
            if transfer_cmp["local_only"]:
                lines.append(f"- Local only: {', '.join(transfer_cmp['local_only'])}")
            if transfer_cmp["amap_only"]:
                lines.append(f"- Amap only: {', '.join(transfer_cmp['amap_only'])}")
            lines.append("")

        lines.append("---")
        lines.append("")
    return lines


def main():
    if len(sys.argv) != 2:
        raise SystemExit("Usage: python -m compare.generate_report <results.json>")

    results_path = sys.argv[1]
    with open(results_path, "r", encoding="utf-8") as f:
        results = json.load(f)

    out_path = os.path.join(os.path.dirname(results_path), "full-report.md")
    with open(out_path, "w", encoding="utf-8") as f:
        f.write("\n".join(build_lines(results)))
    print(f"Report written to: {out_path}")


if __name__ == "__main__":
    main()
